lib/solidus_support/legacy_event_compat/subscriber.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+begin
+  require "omnes"
+rescue LoadError
+end
+
+module SolidusSupport
+  module LegacyEventCompat
+    # Compatibility for subscriber modules
+    #
+    # Thanks to this module, extensions can create legacy subscriber modules
+    # (see {Spree::Event::Subscriber}) and translate them automatically to an
+    # {Omnes::Subscriber}). E.g.:
+    #
+    # ```
+    # module MyExtension
+    #   module MySubscriber
+    #     include Spree::Event::Subscriber
+    #     include SolidusSupport::LegacyEventCompat::Subscriber
+    #
+    #     event_action :order_finalized
+    #
+    #     def order_finalized(event)
+    #       event.payload[:order].do_something
+    #     end
+    #   end
+    # end
+    #
+    # MyExtension::MySubscriber.omnes_subscriber.subscribe_to(Spree::Bus)
+    # ```
+    #
+    # The generated omnes subscriptions will call the corresponding legacy
+    # subscriber method with the omnes event. It'll compatible as long as the
+    # omnes event responds to the `#payload` method (see
+    # {Omnes::UnstructuredEvent}).
+    module Subscriber
+      # @api private
+      ADAPTER = lambda do |legacy_subscriber, legacy_subscriber_method, _omnes_subscriber, omnes_event|
+        legacy_subscriber.send(legacy_subscriber_method, omnes_event)
+      end
+
+      def self.included(legacy_subscriber)
+        legacy_subscriber.define_singleton_method(:omnes_subscriber) do
+          @omnes_subscriber ||= Class.new.include(::Omnes::Subscriber).tap do |subscriber|
+            legacy_subscriber.event_actions.each do |(legacy_subscriber_method, event_name)|
+              subscriber.handle(event_name.to_sym, with: ADAPTER.curry[legacy_subscriber, legacy_subscriber_method])
```

Comment:
Does Omnes pass the `omnes_subscriber` as the first argument? It seems like it [either passes the `event` or the `event` and the `publication_context`](https://github.com/nebulab/omnes/blob/main/lib/omnes/subscription.rb#L56), but I might be mis-interpreting what's going on here.

Reply:
You're right. The subscription callback takes what you said. However, what `with:` needs to return is a callable taking the omnes subscriber as the first argument, while the following two arguments do match subscription callback expectations. That way, the adapters can do their job with the first argument and return a regular subscription. See https://github.com/nebulab/omnes#custom-adapters for details.